### PR TITLE
enhance: use more polyfills and modernize them

### PR DESCRIPTION
**Pull request by marcelgerber**
Inspired by the [recent problem we had with TS & `replaceAll`](https://owid.slack.com/archives/CQQUA2C2U/p1675960336299119).

We were using "relatively modern" features like `.matchAll()` that weren't polyfilled, which was holding us back in terms of browser support when it's easy to polyfill them.

Also, I ran a bunch of browser tests on that branch to get a baseline of which browser we currently support well. Might be useful in the future.

### CODE REVIEW

File: site/SiteConstants.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// See https://polyfill.io/v3/url-builder/ for a list of all supported features
+const polyfillFeatures = [
```
Comment (marcelgerber) [inline site/SiteConstants.ts:2]: Rationale for dropping `es6, fetch, URL`: These are fully supported by all browsers that support async functions, so if the user should use an older browser, nothing will work anyway.

File: site/SiteConstants.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    "es2019", // Array.flat, Array.flatMap, Object.fromEntries, ...
+    "es2020", // String.matchAll, Promise.allSettled, ...
+    "es2021", // String.replaceAll, Promise.any, ...
+    "es2022", // Array.at, String.at, ...
```
Comment (marcelgerber) [inline site/SiteConstants.ts:6]: We don't currently use these, but I found it nice to include them already.
Most users are using modern browsers, where the polyfill will come back empty anyways. So there's not much of an overhead.

File: docs/browser-support.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+### "Most breaking" features
+
+"Most breaking" features we use are:
+
+-   [Async functions](https://caniuse.com/async-functions), which introduce the keywords `async` and `await`.
+-   [Object rest destructuring](https://caniuse.com/mdn-javascript_operators_destructuring_rest_in_objects), which introduces expressions like `const {a, ...rest} = obj`.
+-   [Passing a record to the URLSearchParams constructor](https://caniuse.com/mdn-api_urlsearchparams_urlsearchparams_record), which we use in URL manipulation.
+
+Things will look a bit odd at times in browser that don't support the [CSS grid `gap` property](https://caniuse.com/mdn-css_properties_gap_grid_context), but things still work just fine.
```
Comment (ikesau) [inline docs/browser-support.md:21]:
```suggestion
Things will look a bit odd at times in browsers that don't support the [CSS grid `gap` property](https://caniuse.com/mdn-css_properties_gap_grid_context), but things still work just fine.
```

**Review by ikesau — APPROVED** [review]
Nice work! Thanks for going into detail and improving the documentation while you were at it. 

⭐⭐⭐⭐⭐

**Comment by danyx23** [comment]
Nice that you looked into this! Looks good and I think we should merge and ship this. I'll start a discussion somewhat related to this about the quality of fallback if JS doesn't work/is disabled in the #design channel in slack.

**Pull request merged**